src/intrinsics/ecma262/Object.js

```diff
@@ -101,36 +136,69 @@ export default function(realm: Realm): NativeFunctionValue {
         }
         keys = frm.$OwnPropertyKeys();
 
-        if (to_must_be_partial) {
-          if (to instanceof AbstractObjectValue && to.values.isTop()) {
-            // We don't know which objects to make partial and making all objects partial is failure in itself
-            AbstractValue.reportIntrospectionError(to);
-            throw new FatalError();
-          } else {
-            // if to has properties, we better remove them because after the temporal call to Object.assign we don't know their values anymore
-            if (to.hasStringOrSymbolProperties()) {
-              // preserve them in a snapshot and add the snapshot to the sources
-              delayedSources.push(to.getSnapshot({ removeProperties: true }));
-            }
+        handleSnapshot(frm, frm_was_partial);
+
+        // c. Repeat for each element nextKey of keys in List order,
+        invariant(frm, "from required");
+        invariant(keys, "keys required");
+        copyKeys(keys, frm, to);
+      };
+
+      const tryAndApplySourceOrRecover = nextSource => {
+        let effects;
+        let savedSuppressDiagnostics = realm.suppressDiagnostics;
+        try {
+          realm.suppressDiagnostics = true;
+          effects = realm.evaluateForEffects(
+            () => {
+              applySource(nextSource);
+              return realm.intrinsics.undefined;
+            },
+            undefined,
+            "tryAndApplySourceOrRecover"
+          );
+        } catch (e) {
+          // ignore null or undefined
+          if (nextSource === realm.intrinsics.null || nextSource === realm.intrinsics.undefined) {
```

Comment:
This should be an invariant. If we got here, something went wrong since this should never throw.